File: modules/v1ce-widget-data/android/src/main/java/app/v1ce/widgetdata/V1CEWidgetDataModule.kt
package app.v1ce.widgetdata
import android.content.Intent
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.runBlocking
import app.v1ce.widget.V1CEWidget
private val android.content.Context.v1ceWidgetStore by preferencesDataStore(name = "v1ce_widget")
class V1CEWidgetDataModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("V1CEWidgetData")
    Function("setSnapshot") { json: String ->
      val context = appContext.reactContext ?: return@Function
      runBlocking { context.v1ceWidgetStore.updateData { it.toMutablePreferences().apply { set(stringPreferencesKey("snapshot"), json) } } }
      runCatching { runBlocking { V1CEWidget().updateAll(context) } }
      context.sendBroadcast(Intent("app.v1ce.WIDGET_UPDATE").setPackage(context.packageName))
    }
    Function("clearSnapshot") {
      val context = appContext.reactContext ?: return@Function
      runBlocking { context.v1ceWidgetStore.updateData { it.toMutablePreferences().apply { remove(stringPreferencesKey("snapshot")) } } }
      runCatching { runBlocking { V1CEWidget().updateAll(context) } }
    }
  }
}

File: modules/v1ce-widget-data/ios/V1CEWidgetDataModule.swift
import ExpoModulesCore
import WidgetKit
public class V1CEWidgetDataModule: Module {
  public func definition() -> ModuleDefinition {
    Name("V1CEWidgetData")
    Function("setSnapshot") { (json: String) in
      guard let data = json.data(using: .utf8) else { return }
      UserDefaults(suiteName: "group.app.v1ce")?.set(data, forKey: "v1ce_widget_profile_v1")
      WidgetCenter.shared.reloadAllTimelines()
    }
    Function("clearSnapshot") {
      UserDefaults(suiteName: "group.app.v1ce")?.removeObject(forKey: "v1ce_widget_profile_v1")
      WidgetCenter.shared.reloadAllTimelines()
    }
  }
}